Configuration class for BatchSpanProcessor

thisthat opened the pull request:
This addresses #996 creating a configuration class for the `BatchSpanProcessor`.
The `ConfigBuilder` class contains the common code that will be shared with other configuration classes.

Oberon00 commented on sdk/src/main/java/io/opentelemetry/sdk/common/ConfigBuilder.java:
```diff
@@ -0,0 +1,52 @@
+
+import java.util.Map;
+
+public abstract class ConfigBuilder {
```
```suggestion
public abstract class ConfigurableBuilder {
```

Looking at the usages, this is not a builder for configuration, but the interface for a configurable builder. Consequently, I think the getProperty methods must be removed from here (they can be moved to a static helper class) and then this can become an interface instead of an abstract class.

jkwatson replied:
Agree 100% with @Oberon00 's approach. We should try not to introduce class hierarchies when we can avoid it.

Oberon00 commented on sdk/src/main/java/io/opentelemetry/sdk/common/ConfigBuilder.java:
```diff
@@ -0,0 +1,52 @@
+
+  public abstract ConfigBuilder fromConfigMap(Map<String, String> configMap);
+
+  public abstract ConfigBuilder fromEnv();
```
Since this is expected to call `fromConfigMap(System.getenv()`, I'd move it out of the interface.

Oberon00 commented on sdk/src/main/java/io/opentelemetry/sdk/common/ConfigBuilder.java:
```diff
@@ -0,0 +1,52 @@
+
+  protected ConfigBuilder() {}
+
+  public abstract ConfigBuilder fromConfigMap(Map<String, String> configMap);
```
What naming conventions do we expect in configMap? `ENV_VAR_CONVENTIONS`? Or `systemProperty.conventions`? It seems the former.

thisthat replied:
From https://github.com/open-telemetry/opentelemetry-java/issues/996#issuecomment-597374377 it is the former, yes.
@bogdandrutu and @trask could you please provide more input on this?

trask replied:
I like the convention of looking for system properties with `one.twoThree` and looking for the same env var with `ONE_TWO_THREE`. Since the translation is not reversible (does `ONE_TWO_THREE` map to `one.two.three` or `oneTwo.three` or ...), I think it makes sense to use the system property names in the internal map.

bogdandrutu commented on sdk/src/main/java/io/opentelemetry/sdk/common/ConfigBuilder.java:
```diff
@@ -0,0 +1,52 @@
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.common;
```
common.export?

bogdandrutu commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +321,184 @@ public void run() {
   }
+
+  @Immutable
+  public static final class Config {
```
Use autovalue to avoid writing unnecessary code?

bogdandrutu commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +321,184 @@ public void run() {
+      private static final String KEY_EXPORT_TIMEOUT_MILLIS = "OTEL_BSP_EXPORT_TIMEOUT";
+      private static final String KEY_SAMPLED = "OTEL_BSP_REPORT_SAMPLED";
+
+      private static final long SCHEDULE_DELAY_MILLIS = 5000;
+      private static final int MAX_QUEUE_SIZE = 2048;
+      private static final int MAX_EXPORT_BATCH_SIZE = 512;
+      private static final int EXPORT_TIMEOUT_MILLIS = 30_000;
+      private static final boolean REPORT_ONLY_SAMPLED = true;
```
It is hard to understand these names, I would propose to prefix all of them with "DEFAULT_"

jkwatson replied:
I'd also like to have them documented outside of these private fields. Maybe put them in a package-info file?

bogdandrutu commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +321,184 @@ public void run() {
+      return new Builder();
+    }
+
+    public static class Builder extends ConfigBuilder {
```
final?

bogdandrutu commented on sdk/src/main/java/io/opentelemetry/sdk/common/ConfigBuilder.java:
```diff
@@ -0,0 +1,52 @@
+
+  public abstract ConfigBuilder fromEnv();
+
+  protected boolean getProperty(String name, Map<String, String> map, boolean defaultValue) {
+    try {
+      return Boolean.parseBoolean(System.getProperty(name, map.get(name)));
+    } catch (NumberFormatException ex) {
+      return defaultValue;
+    }
+  }
+
+  protected int getProperty(String name, Map<String, String> map, int defaultValue) {
+    try {
+      return Integer.parseInt(System.getProperty(name, map.get(name)));
+    } catch (NumberFormatException ex) {
+      return defaultValue;
+    }
+  }
+
+  protected long getProperty(String name, Map<String, String> map, long defaultValue) {
+    try {
+      return Long.parseLong(System.getProperty(name, map.get(name)));
+    } catch (NumberFormatException ex) {
+      return defaultValue;
+    }
+  }
```
I would not hardcode that we always look in System properties then env then default. I would allow the following workflow:
* Start from default;
* Apply changes from env;
* Apply changes from System properties;

```
BatchSpanProcessor.getBuilder().overwriteFromEnv().overwriteFromSystemProperties().build()
```

The default should be the proposed solution but would like if the users can also change this order if needed.

Thoughts on this?

Oberon00 replied:
I'd even prefer completely separating thes steps of collection and application of configuration. The builder just has `applyFromMap` and the collection happens separately.

I.e., I propose something like:

```java
final Config cfg = Config.newBuilder().readEnv().readSystemProps().readCommandLine(args).set(MY_KEY, MY_OVERRIDE_VALUE).build();
final BatchSpanProcessor processor = BatchSpanProcessor.newBuilder().applyConfig(cfg).build();
```

Oberon00 replied:
If we allowed `Map<String, Object>` (or stored that in the hypothetical Configuration class) we could even support different representations of complex objects like resources from different sources, e.g. JSON object in JSON, URL/query encoded in envvar, ....

bogdandrutu replied:
I liked the proposed API to build the config. "nit: applyConfig -> setConfig" probably better name for a builder.

Also I expect that "readX" methods only apply changes to arguments defined by "X" correct? Does not overwrite all the arguments with default if not defined in the config.

We can also offer a helper `Config.loadFromDefaultSources()` (we can think of a better name) which essentially allows users to avoid the 5 lines you wrote to build the config.

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +321,184 @@ public void run() {
+      private boolean sampled = REPORT_ONLY_SAMPLED;
+
+      /**
+       * Sets the configuration values from the given configuration map.
```
Add a comment about the keys and what values are expected for them.

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +321,184 @@ public void run() {
+      }
+
+      /**
+       * Sets the configuration values from environment variables.
```
Add a link to the relevant env vars?

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +321,184 @@ public void run() {
+        return this;
+      }
+
+      public Config build() {
```
javadoc needed here.

jkwatson commented:
This PR proposes replacing Builders with Config classes. Any idea what set of SDK entities would end up being converted over to Config-based creation?

bogdandrutu commented:
@jkwatson I think I posted in the issue what is the target

bogdandrutu commented:
@thisthat any progress on this? I think we added some ideas in the comments.

thisthat commented:
Sorry, I was pretty busy this week with internal stuff. I will work on this in the next days!

thisthat commented:
After 957ecc2, I have two questions:

1. Is it ok to normalize with lowercase and dot separation the lookup of variables? Currently, `MY_ENV_VAR` is the same as `my.env.var` which is the same as `MY_env.VaR`.

2. Since the configuration for other classes will look like the same, do we want to have a single configuration class that holds all configuration parameters of OTel?

carlosalberto commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+      private int exporterTimeoutMillis = DEFAULT_EXPORT_TIMEOUT_MILLIS;
+      private boolean sampled = DEFAULT_REPORT_ONLY_SAMPLED;
+
+      enum NamingConvention {
```
So I feel like this could be re-used by other configuration logic. But of course, lets do that in a follow-up PR ;)

carlosalberto commented:
> Not an expert of Zipkin, but overall looks great @jkwatson You need to add the new artifact to `all/build.gradle` too, btw ;)

I'd go with YES.

> Since the configuration for other classes will look like the same, do we want to have a single configuration class that holds all configuration parameters of OTel?

Strong yes here - I think lots of these naming convention/config routines can be reused in other places. It can be done in a follow-up PR though.

jkwatson commented:
> > Not an expert of Zipkin, but overall looks great @jkwatson You need to add the new artifact to `all/build.gradle` too, btw ;)

Not sure why this comment ended up over here, @carlosalberto

bogdandrutu commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+
+        public Map<String, String> normalize(@Nonnull Map<String, String> map) {
+          Map<String, String> properties = new HashMap<>();
+          for (String key : map.keySet()) {
+            properties.put(normalize(key), map.get(key));
+          }
```
Iterate over entries not keyset if you need access to the values as well.

bogdandrutu commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+       * @param configMap {@link Map} holding the configuration values.
+       * @return this.
+       */
+      public Builder fromConfigMap(
```
Do we need to have this public?

jkwatson replied:
If someone wanted to use this to build config from a parsed yaml/json/xml file, this would be a handy entry point.

bogdandrutu replied:
Not really, you cannot extend the enum, so it will support only DOT or ENV for the enum. Also I feel like we should not expose helpers just in case someone needs it (we will suffer maintaining this API). I would suggest to remove it.

thisthat replied:
I left it public because I thought that the agent might load the json/xml/yaml file and use this API to set the configuration object.
So the agent will properly create the map with the ENV or DOT naming convention.

thisthat replied:
In the SIG call, we agreed on supporting Properties as an additional configuration medium. This method can be package visible only.

bogdandrutu commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+      return new Builder();
+    }
+
+    public static final class Builder {
```
We can use auto-value for this as well to remove a lot of variables end ensure no bugs coming from copy-paste.

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+            return key.toLowerCase();
+          }
+        },
+        ENV_VAR {
```
was just going to say the same. we should normalize the env vars to uppercase.

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/package-info.java:
```diff
@@ -0,0 +1,40 @@
+ *   <li>{@link io.opentelemetry.sdk.trace.export.MultiSpanExporter}
+ * </ul>
+ *
+ * <h2>Default values for {@link io.opentelemetry.sdk.trace.export.BatchSpansProcessor}</h2>
```
This document should specify both the env vars, the system properties, and the defaults, I think.

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+            sampled, scheduleDelayMillis, maxQueueSize, maxExportBatchSize, exporterTimeoutMillis);
+      }
+    }
+  }
```
this class is now over 600 lines long. I think it might be time to break out the Config and the Builder into their own files.

jkwatson replied:
not sure why you resolved this with no response, @thisthat

thisthat replied:
Sorry, I forgot to comment on the decision taken in the SIG call.
Refactoring will be done in a separate PR when the configuration for other parts of OTel will be provided.

jkwatson replied:
ah, ok. I didn't know that that would be a part of that effort.  👍

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+        }
+      }
+
+      private static long getProperty(String name, Map<String, String> map, long defaultValue) {
```
I'd prefer these methods were named `getLongProperty`, `getIntProperty`, etc.

jkwatson commented:
Is there a desire to have unified env vars across language implementations?  If so, should we write a spec for these, for more widespread review?

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+      /**
+       * Sets the maximum time an exporter will be allowed to run before being cancelled.
+       *
+       * <p>Default value is {@code 30000}ms
```
can we add a `@see` that links to the constant, just in case things get out of synch? Also for the other properties.

jkwatson commented on sdk/src/test/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessorTest.java:
```diff
@@ -73,6 +75,116 @@ private ReadableSpan createSampledEndedSpan(String spanName) {
   }
 
+  @Test
+  public void testConfiguration() {
```
Is there any way we can get actually pulling the config from env vars/system properties? I know that's generally a pain, but I don't know if someone had some prior art on making that kind of test to verify that this works for real.

thisthat replied:
Unfortunately, I don't have prior art. But I will search for a proper solution!

trask replied:
I've used this before: https://stefanbirkner.github.io/system-rules

thisthat replied:
Thank you @trask 👍

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+    /**
+     * Creates a {@link Config} object reading the configuration values first from the environment
+     * and then from system properties. If a configuration value is missing, it uses the default
+     * value.
```
Can we add a wee bit of clarification that the system property will override the environment values?

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+            getProperty(KEY_SCHEDULE_DELAY_MILLIS, configMap, this.scheduleDelayMillis));
+        this.setMaxQueueSize(
+            getProperty(
+                namingConvention.normalize(KEY_MAX_QUEUE_SIZE), configMap, this.maxQueueSize));
```
extra normalize here?

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+       *
+       * @return this.
+       */
+      public Builder readEnv() {
```
how about renaming to `readFromEnvironment` or `readEnvironment`

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+
+    public abstract int getExporterTimeoutMillis();
+
+    public static Config getDefault() {
```
javadoc needed here.

jkwatson commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +324,277 @@ public void run() {
+  @AutoValue
+  public abstract static class Config {
+
+    public abstract boolean isReportOnlySampled();
```
how about `exportOnlySampled()` as the name for this? `report` isn't a term that shows up anywhere else that I know of.

thisthat replied:
I also found that strange when I look at the class the first time! Totally in favor of `exportOnlySampled()`.

bogdandrutu commented:
@thisthat this PR is on you now, we need fixes for the round of comments we have for the moment.

thisthat commented:
> Is there a desire to have unified env vars across language implementations? If so, should we write a spec for these, for more widespread review?

I will open an issue in the spec after we are done with the configuration for all classes so I can provide the list of env vars we need. Regarding this topic, there is already an issue for setting the exporter host: https://github.com/open-telemetry/opentelemetry-specification/issues/172

@bogdandrutu and @jkwatson, I have addressed your the suggestions, PTAL.

bogdandrutu commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +327,372 @@ public void run() {
+     * @return a new {@code Builder} with default options.
+     * @since 0.4.0
+     */
+    public static Builder builder() {
```
I think we use the convention `newBuilder` for this case.

bogdandrutu commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +327,372 @@ public void run() {
+          }
+        };
+
+        @Nonnull
```
We don't use the `@Nonnull` convention. Please avoid adding this new convention yet.

bogdandrutu commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +327,372 @@ public void run() {
+       * @param properties {@link Properties} holding the configuration values.
+       * @return this.
+       */
+      public Builder readProperties(Properties properties) {
+        Long longValue = getLongProperty(KEY_SCHEDULE_DELAY_MILLIS, properties);
+        if (longValue != null) {
+          this.setScheduleDelayMillis(longValue);
+        }
+        Integer intValue = getIntProperty(KEY_MAX_QUEUE_SIZE, properties);
+        if (intValue != null) {
+          this.setMaxQueueSize(intValue);
+        }
+        intValue = getIntProperty(KEY_MAX_EXPORT_BATCH_SIZE, properties);
+        if (intValue != null) {
+          this.setMaxExportBatchSize(intValue);
+        }
+        intValue = getIntProperty(KEY_EXPORT_TIMEOUT_MILLIS, properties);
+        if (intValue != null) {
+          this.setExporterTimeoutMillis(intValue);
+        }
+        Boolean boolValue = getBooleanProperty(KEY_SAMPLED, properties);
+        if (boolValue != null) {
+          this.setExportOnlySampled(boolValue);
+        }
+        return this;
+      }
```
Why do you not call `fromConfigMap`? Seems a lot of duplicated code. Do I miss anything? If yes maybe a comment would help others to also understand.

bogdandrutu replied:
`return fromConfigMap(Maps.fromProperties(properties), NamingConvention.DOT);`

thisthat replied:
I overlooked the possibility to coerce a Property object into a Map!

bogdandrutu commented on sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java:
```diff
@@ -411,4 +327,372 @@ public void run() {
+       * @return this.
+       */
+      public Builder readSystemProperties() {
+        return fromConfigMap(Maps.fromProperties(System.getProperties()), NamingConvention.DOT);
```
Would call `return readProperties(System.getProperties())`

jkwatson commented:
> Is there a desire to have unified env vars across language implementations? If so, should we write a spec for these, for more widespread review?

See https://github.com/open-telemetry/opentelemetry-specification/issues/572

carlosalberto commented:
As this PR looks ready to be merged: shall we create an issue to keep track of the upcoming configuration refactoring, so we can use this all over our SDK components?

jkwatson commented:
> As this PR looks ready to be merged: shall we create an issue to keep track of the upcoming configuration refactoring, so we can use this all over our SDK components?

I think the refactoring should come naturally by adding the rest of the configuration. I don't think we need a separate issue to track it.

bogdandrutu commented:
@jkwatson what about my comments? Do we want to get them fixed?

jkwatson commented:
> @jkwatson what about my comments? Do we want to get them fixed?

Sure. I thought that since you had approved, that they had all been resolved. @thisthat could you respond to the last couple of @bogdandrutu 's comments?

thisthat commented:
> > @jkwatson what about my comments? Do we want to get them fixed?
> 
> Sure. I thought that since you had approved, that they had all been resolved. @thisthat could you respond to the last couple of @bogdandrutu 's comments?

I totally agree with @bogdandrutu 's comment and I have addressed them!
